e2e/launcher/pocketserver.go

```diff
@@ -0,0 +1,124 @@
+package launcher
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+	"sync"
+)
+
+type PocketServer interface {
+	Start(...string) error
+	Kill() error
+	RegisterPatternActor(patternActor PatternActor, stream Stream) error
+}
+
+func NewPocketServer(executableLocation string) PocketServer {
+	return &pocketServer{
+		executableLocation: executableLocation,
+		pocketInstance:     nil,
+	}
+}
+
+type pocketServer struct {
+	executableLocation string
+	pocketInstance     *exec.Cmd
+	stdOutPipeline     *PatternActionPipeline
+	stdErrPipeline     *PatternActionPipeline
+}
+
+func (ps *pocketServer) Start(arguments ...string) error {
+	if ps.pocketInstance != nil {
+		return errors.New("pocket instance already started")
+	}
+	invocation := []string{"start"}
+	invocation = append(invocation, arguments...)
+	ps.pocketInstance = exec.Command(ps.executableLocation, invocation...)
+	ps.stdOutPipeline = &PatternActionPipeline{patternActors: []PatternActor{}}
+	ps.stdErrPipeline = &PatternActionPipeline{patternActors: []PatternActor{}}
+	ps.pocketInstance.Stdout = ps.stdOutPipeline
+	ps.pocketInstance.Stderr = ps.stdErrPipeline
+	err := ps.pocketInstance.Start()
+	return err
+}
+
+func (ps *pocketServer) Kill() error {
+	return ps.pocketInstance.Process.Kill()
+}
+
+type Stream int
+
+// matching standard file descriptors
+const (
+	StdIn  = Stream(0)
+	StdOut = Stream(1)
+	StdErr = Stream(2)
+)
+
+func (ps *pocketServer) RegisterPatternActor(patternActor PatternActor, stream Stream) error {
+	switch stream {
+	case StdOut:
+		ps.stdOutPipeline.patternActors = append(ps.stdOutPipeline.patternActors, patternActor)
+	case StdErr:
+		ps.stdErrPipeline.patternActors = append(ps.stdOutPipeline.patternActors, patternActor)
+	default:
+		return errors.New("invalid stream")
+	}
+	return nil
+}
+
+type PatternActor interface {
```

Comment:
DISCUSS: I've always been a fan of declaring all the types in one place or at the top of the file.

I've never gotten accustomed to having types be defined midfile, but is this a common golang pattern?

Reply:
My pattern is shaped by years of Java development (declaration and definition are together, so types end up mid-file). The pattern you prefer reminds me of C and C++ development, where declaration and definition are separate. 

Navigation between files is usually done by symbol cross reference, so we don't waste much time finding things because they're far from the top; furthermore, having all of the types at the top will yield a group of functions unrelated to each other hanging out together at the end of the file.

I had way more text, but decided to go to `io/io.go` in the standard library for guidance. This is how that file is ordered:
- Package level constants and variables
- Interfaces
- Functions with no receivers (operate on interfaces)
- all the structs with their related data around:
  - struct initializer (NewXxYy and similar)
  - struct
  - functions with the struct or pointer to the struct as a receiver

structs are ordered so that the ones below depend on the ones above.

I think this works because of this:
- First you see package level things are readily available and you might need to customize/refer to
- Then interfaces you might want to implement
- Then functions you might need to use on the interfaces
- Then structs related things near their structs
- In the only case you'd read from top to bottom (generally when studying a module to learn it/from it) the fact that types go from more general to more specific means that you get a sense for the architecture / rationale.

I think we should emulate this pattern in our codebases.

Reply:
Sounds good to me. Leave as is.